p2panda-rs/src/secret_group/group.rs

```diff
@@ -0,0 +1,552 @@
+        data: &[u8],
+    ) -> Result<SecretGroupMessage, SecretGroupError> {
+        let mls_ciphertext = self.mls_group.encrypt(provider, data)?;
+        Ok(SecretGroupMessage::MlsApplicationMessage(mls_ciphertext))
+    }
+
+    /// Encrypt user data using the group's current symmetrical long term secret.
+    ///
+    /// This method gives only post-compromise security and has in general lower security
+    /// guarantees but gives more flexibility. Use this encryption method if you want every old or
+    /// new group member to decrypt past data even when they've joined the group later.
+    pub fn encrypt_with_long_term_secret(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        data: &[u8],
+    ) -> Result<SecretGroupMessage, SecretGroupError> {
+        // Generate unique nonce for AEAD encryption
+        let nonce = self.generate_nonce(provider)?;
+
+        // Unwrap here since at this stage we already have at least one LTS epoch
+        let epoch = self.long_term_epoch().unwrap();
+        let secret = self.long_term_secret(epoch).unwrap();
+
+        // Encrypt user data with last long term secret
+        let ciphertext = secret.encrypt(provider, &nonce, data)?;
+
+        Ok(SecretGroupMessage::LongTermSecretMessage(ciphertext))
+    }
+
+    /// Decrypts user data.
+    ///
+    /// This method automatically detects if the ciphertext was encrypted with a Sender Ratchet
+    /// Secret or a Long Term Secret and returns an `SecretGroupError` if the required key material
+    /// for decryption is missing.
+    pub fn decrypt(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        message: &SecretGroupMessage,
+    ) -> Result<Vec<u8>, SecretGroupError> {
+        match message {
+            SecretGroupMessage::MlsApplicationMessage(ciphertext) => {
+                Ok(self.mls_group.decrypt(provider, ciphertext.clone())?)
+            }
+            SecretGroupMessage::LongTermSecretMessage(ciphertext) => {
+                let secret = self
+                    .long_term_secret(ciphertext.long_term_epoch())
+                    .ok_or_else(|| SecretGroupError::LTSSecretMissing)?;
+                Ok(secret.decrypt(provider, ciphertext)?)
+            }
+        }
+    }
+
+    // Status
+    // ======
+
+    /// Returns true if this group is still active or false if the member got removed from the
+    /// group.
+    pub fn is_active(&self) -> bool {
+        self.mls_group.is_active()
+    }
+
+    /// Returns true if this group is owned by us.
+    pub fn is_owned(&self) -> bool {
+        self.owned
+    }
+
+    /// Returns the hash of this `SecretGroup` instance.
+    pub fn group_instance_id(&self) -> Hash {
+        let group_id_bytes = self.mls_group.group_id().as_slice().to_vec();
+        // Unwrap here since we already trusted the user input
+        Hash::new_from_bytes(group_id_bytes).unwrap()
+    }
+
+    /// Returns the current epoch of the long term secret.
```

Comment:
When a group got freshly created! I could rewrite the code to start with Epoch `0` and increment it AFTER the rotation of the secret, maybe this makes it easier to read?

Reply:
I think it would be enough to add this info to the docstring so when somebody uses the `long_term_epoch` method they know how to handle the `None` case.